subsys/lorawan/services/frag_decoder_lowmem.c

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright (c) 2024 A Labs GmbH
+ * Copyright (c) 2024 tado GmbH
+ * Copyright (c) 2022 Jiapeng Li
+ *
+ * Based on: https://github.com/JiapengLi/LoRaWANFragmentedDataBlockTransportAlgorithm
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "frag_decoder_lowmem.h"
+#include "frag_flash.h"
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/bitarray.h>
+
+LOG_MODULE_REGISTER(lorawan_frag_dec, CONFIG_LORAWAN_SERVICES_LOG_LEVEL);
+
+SYS_BITARRAY_DEFINE_STATIC(lost_frames, FRAG_MAX_NB);
+SYS_BITARRAY_DEFINE_STATIC(lost_frames_matrix, (FRAG_TOLERANCE * (FRAG_TOLERANCE + 1) / 2));
+SYS_BITARRAY_DEFINE_STATIC(matched_lost_frm_bm0, FRAG_TOLERANCE);
+SYS_BITARRAY_DEFINE_STATIC(matched_lost_frm_bm1, FRAG_TOLERANCE);
+SYS_BITARRAY_DEFINE_STATIC(matrix_line_bm, FRAG_MAX_NB);
+
+static inline size_t matrix_location_to_index(size_t x, size_t y, size_t m);
```

Comment:
can all of these forward declarations be avoided if `frag_dec` is moved to the end of the file?

Reply:
Yes, no problem. I did the forward declaration stuff because I wanted the most important implementation first in the file, but that's a matter of taste and I don't mind changing it.